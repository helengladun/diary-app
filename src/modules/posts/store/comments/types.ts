export enum PostCommentsTypes {
    GET_POST_COMMENTS = '@@posts/GET_POST_COMMENTS',
    GET_POST_COMMENTS_SUCCESS = '@@posts/GET_POST_COMMENTS_SUCCESS',
    GET_POST_COMMENTS_FAILURE = '@@posts/GET_POST_COMMENTS_FAILURE',

    CLEAR_POST_COMMENTS = '@@posts/CLEAR_POST_COMMENTS',
}