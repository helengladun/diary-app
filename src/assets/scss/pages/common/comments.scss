.comments {
  max-width: 444px;
  width: 100%;
  padding: 8px 19px 14px 18px;

  border: 1px solid $primaryBorderColor;
  background-color: $primaryLightColor;
  box-sizing: border-box;

  &__content {
    .comment:last-child {
      border-bottom: none;
    }
  }

  &__title {
    margin-bottom: 20px;
    text-align: left;
  }
}