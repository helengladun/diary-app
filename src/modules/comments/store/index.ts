import {combineReducers} from 'redux';
import {ICommentsState, reducer as commentsReducer} from "./all-comments";

export interface CommentsState {
    allComments: ICommentsState,
}

export const reducer = combineReducers<CommentsState>({
    allComments: commentsReducer,
});
