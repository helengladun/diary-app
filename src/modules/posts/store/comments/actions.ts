import {action} from "typesafe-actions";
import {PostCommentsTypes} from "./types";
import {IComment} from "../../../shared/models/IComment";

export const getPostComments = () => action(PostCommentsTypes.GET_POST_COMMENTS);
export const getPostCommentsSuccess = (data: IComment[]) => action(PostCommentsTypes.GET_POST_COMMENTS_SUCCESS, {data});
export const getPostCommentsFailure = (message: string) => action(PostCommentsTypes.GET_POST_COMMENTS_FAILURE, { message });

export const clearPostComments = () => action(PostCommentsTypes.CLEAR_POST_COMMENTS);
