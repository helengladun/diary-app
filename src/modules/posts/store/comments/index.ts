import { AnyAction, Reducer } from 'redux';
import { PostCommentsTypes } from './types';
import {merge} from 'ramda';
import {IComment} from "../../../shared/models/IComment";

export interface IPostCommentsState {
    entity: IComment[],
    pending: boolean,
    message: string,
}

const initialState: IPostCommentsState = {
    pending: false,
    entity: [],
    message: ''
};

export const reducer: Reducer<IPostCommentsState> = (state = initialState, action: AnyAction) => {
    switch (action.type) {
        case PostCommentsTypes.GET_POST_COMMENTS:
            return merge(state, {
                pending: true
            });
        case PostCommentsTypes.GET_POST_COMMENTS_SUCCESS:
            return merge(state, {
                pending: false,
                entity: action.payload.data
            });
        case PostCommentsTypes.GET_POST_COMMENTS_FAILURE:
            return merge(state, {
                pending: false,
                message: action.payload.message
            });
        case PostCommentsTypes.CLEAR_POST_COMMENTS:
            return merge(state, {
                pending: false,
                entity: []
            });
        default:
            return state;
    }
};