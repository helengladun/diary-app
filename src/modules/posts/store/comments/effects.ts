import {ThunkDispatch} from "redux-thunk";
import {IApplicationState} from "../../../store/rootReducer";
import {Action} from "redux";
import {getPostComments, getPostCommentsFailure, getPostCommentsSuccess } from "./actions";
import {PostService} from "../../../shared/services/PostService";

export const getPostCommentsEffect = (post_id: number) => (dispatch: ThunkDispatch<IApplicationState, void, Action>, getState: Function) => {
    dispatch(getPostComments());

    try {
        const res = PostService.getComments(post_id);
        dispatch(getPostCommentsSuccess(res));
    } catch (e) {
        const message = 'Error found. Please, try again.';
        dispatch(getPostCommentsFailure(message));
    }
};