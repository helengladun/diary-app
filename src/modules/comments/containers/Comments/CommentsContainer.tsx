import React, {ChangeEvent, Component, KeyboardEvent} from 'react';
import {isEmpty} from "ramda";
import {IComment} from "../../../shared/models/IComment";
import {AddCommentForm} from "../../components/AddCommentForm/AddCommentForm";
import {Comment} from "../../components/Comment/Comment";
import {addCommentEffect} from "../../store/all-comments/effects";
import {connect} from "react-redux";
import {clearPostComments} from "../../../posts/store/comments/actions";
import uuid from 'uuid';

interface PropsFromDispatch {
    addCommentEffect: Function,
    clearPostComments: Function
}

interface IProps extends PropsFromDispatch {
    data: IComment[],
    postId: number
}

interface IState {
    commentText: string,
}

export class CommentsContainer extends Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);

        this.state = {
            commentText: '',
        };

        this.changeCommentHandler = this.changeCommentHandler.bind(this);
        this.submitCommentFormHandler = this.submitCommentFormHandler.bind(this);
        this.clearCommentValue = this.clearCommentValue.bind(this);
    }

    componentWillUnmount(): void {
        this.props.clearPostComments()
    }

    changeCommentHandler(e: ChangeEvent<HTMLTextAreaElement>) {
        const text = e.target.value;
        this.setState(() => ({commentText: text}));
    }

    clearCommentValue() {
        this.setState(() => ({commentText: ''}));
    }

    submitCommentFormHandler(e: KeyboardEvent<HTMLTextAreaElement>) {
        e.preventDefault();

        if (e.which === 13 && e.ctrlKey) {
            this.props.addCommentEffect({
                id: uuid(),
                description: this.state.commentText,
                post_id: this.props.postId
            });

            this.clearCommentValue();
        }
    }
    
    render() {
        return (
            <div className="comments">
                <h2 className="comments__title title title_1">Comments #{this.props.postId}</h2>
                {!isEmpty(this.props.data) && this.props.data && this.getCommentsList(this.props.data)}
                <AddCommentForm changeHandler={this.changeCommentHandler}
                                value={this.state.commentText}
                                submitHandler={this.submitCommentFormHandler}/>
            </div>
        );
    }

    private getCommentsList(items: IComment[]) {
        return (
            <div className="comments__content">
                {items.map((item: IComment) => (
                    <Comment key={item.id} data={item}/>
                ))}
            </div>
        )
    }
}

const mapDispatchToProps: PropsFromDispatch = {
    addCommentEffect,
    clearPostComments
};

export default connect(null, mapDispatchToProps)(CommentsContainer);