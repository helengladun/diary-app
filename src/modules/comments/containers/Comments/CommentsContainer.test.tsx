import React from 'react';
import {shallow} from 'enzyme';
import {CommentsContainer} from "./CommentsContainer";
import {comments} from "../../../shared/fixtures/comments";

describe('CommentsContainer test', () => {

    const addCommentEffect = jest.fn();
    const clearPostComments = jest.fn();
    const wrapper = shallow(<CommentsContainer
        addCommentEffect={addCommentEffect}
        clearPostComments={clearPostComments}
        data={comments}
        postId={2}
    />);

    test('should render CommentsContainer correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
});