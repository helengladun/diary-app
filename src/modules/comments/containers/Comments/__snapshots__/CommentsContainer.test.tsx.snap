// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsContainer test should render CommentsContainer correctly 1`] = `
<div
  className="comments"
>
  <h2
    className="comments__title title title_1"
  >
    Comments #
    2
  </h2>
  <div
    className="comments__content"
  >
    <Comment
      data={
        Object {
          "description": "Cum habena peregrinatione, omnes parses anhelare flavum, placidus brabeutaes.",
          "id": 1,
          "post_id": 2,
        }
      }
      key="1"
    />
    <Comment
      data={
        Object {
          "description": "Adgium persuaderes, tanquam raptus cacula. Demissios ire! Solitudo de pius valebat, experientia orexis!",
          "id": 2,
          "post_id": 2,
        }
      }
      key="2"
    />
    <Comment
      data={
        Object {
          "description": "Cobaltums mori in pius hafnia!",
          "id": 3,
          "post_id": 1,
        }
      }
      key="3"
    />
  </div>
  <AddCommentForm
    changeHandler={[Function]}
    submitHandler={[Function]}
    value=""
  />
</div>
`;
